Properly reject Iceberg tables in Hive connector

Currently 

```
SELECT * FROM hive.default.an_iceberg_table
```

fails with

```
Query 20211207_145628_00264_6i3st failed: Unable to create input format org.apache.hadoop.mapred.FileInputFormat
io.trino.spi.TrinoException: Unable to create input format org.apache.hadoop.mapred.FileInputFormat
	at io.trino.plugin.hive.util.HiveUtil.getInputFormat(HiveUtil.java:342)
	at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadPartition(BackgroundHiveSplitLoader.java:388)
	at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadSplits(BackgroundHiveSplitLoader.java:352)
	at io.trino.plugin.hive.BackgroundHiveSplitLoader$HiveSplitLoaderTask.process(BackgroundHiveSplitLoader.java:276)
	at io.trino.plugin.hive.util.ResumableTasks$1.run(ResumableTasks.java:38)
	at io.trino.$gen.Trino_dev____20211207_122558_2.run(Unknown Source)
	at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:80)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: java.lang.InstantiationException
	at org.apache.hadoop.util.ReflectionUtils.newInstance(ReflectionUtils.java:135)
	at io.trino.plugin.hive.util.HiveUtil.getInputFormat(HiveUtil.java:339)
	... 9 more
Caused by: java.lang.InstantiationException
	at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.hadoop.util.ReflectionUtils.newInstance(ReflectionUtils.java:133)
	... 10 more
```

and

```
SELECT * FROM hive.default."an_iceberg_table$properties"
```

does not fail at all.

--- a/plugin/trino-hive-hadoop2/src/test/java/io/trino/plugin/hive/TestHiveAlluxioMetastore.java
+++ b/plugin/trino-hive-hadoop2/src/test/java/io/trino/plugin/hive/TestHiveAlluxioMetastore.java
@@ -167,6 +167,13 @@ public class TestHiveAlluxioMetastore
     }
 
     @Override
+    public void testDisallowQueryingOfIcebergTables()
+    {
+        // Alluxio metastore does not support create operations
+        throw new SkipException("not supported");
+    }
+
+    @Override
     public void testIllegalStorageFormatDuringTableScan()
     {
         // Alluxio metastore does not support create operations

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java
@@ -267,6 +267,7 @@ import static io.trino.plugin.hive.util.HiveUtil.getRegularColumnHandles;
 import static io.trino.plugin.hive.util.HiveUtil.hiveColumnHandles;
 import static io.trino.plugin.hive.util.HiveUtil.isDeltaLakeTable;
 import static io.trino.plugin.hive.util.HiveUtil.isHiveSystemSchema;
+import static io.trino.plugin.hive.util.HiveUtil.isIcebergTable;
 import static io.trino.plugin.hive.util.HiveUtil.toPartitionValues;
 import static io.trino.plugin.hive.util.HiveUtil.verifyPartitionTypeSupported;
 import static io.trino.plugin.hive.util.HiveWriteUtils.checkTableIsWritable;
@@ -435,6 +436,9 @@ public class HiveMetadata
 
         if (isDeltaLakeTable(table)) {
             throw new TrinoException(HIVE_UNSUPPORTED_FORMAT, format("Cannot query Delta Lake table '%s'", tableName));
+        }
+        if (isIcebergTable(table)) {
+            throw new TrinoException(HIVE_UNSUPPORTED_FORMAT, format("Cannot query Iceberg table '%s'", tableName));
         }
 
         // we must not allow system tables due to how permissions are checked in SystemTableAwareAccessControl

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PartitionsSystemTableProvider.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PartitionsSystemTableProvider.java
@@ -43,6 +43,7 @@ import static io.trino.plugin.hive.util.HiveBucketing.getHiveBucketHandle;
 import static io.trino.plugin.hive.util.HiveUtil.getPartitionKeyColumnHandles;
 import static io.trino.plugin.hive.util.HiveUtil.getRegularColumnHandles;
 import static io.trino.plugin.hive.util.HiveUtil.isDeltaLakeTable;
+import static io.trino.plugin.hive.util.HiveUtil.isIcebergTable;
 import static io.trino.plugin.hive.util.SystemTables.createSystemTable;
 import static java.util.Objects.requireNonNull;
 import static java.util.function.Function.identity;
@@ -82,7 +83,7 @@ public class PartitionsSystemTableProvider
         Table sourceTable = metadata.getMetastore()
                 .getTable(new HiveIdentity(session), sourceTableName.getSchemaName(), sourceTableName.getTableName())
                 .orElse(null);
-        if (sourceTable == null || isDeltaLakeTable(sourceTable)) {
+        if (sourceTable == null || isDeltaLakeTable(sourceTable) || isIcebergTable(sourceTable)) {
             return Optional.empty();
         }
         verifyOnline(sourceTableName, Optional.empty(), getProtectMode(sourceTable), sourceTable.getParameters());

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PartitionsSystemTableProvider.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PartitionsSystemTableProvider.java
@@ -14,6 +14,7 @@
 package io.trino.plugin.hive;
 
 import io.trino.plugin.hive.authentication.HiveIdentity;
+import io.trino.plugin.hive.metastore.Table;
 import io.trino.spi.connector.ColumnMetadata;
 import io.trino.spi.connector.ConnectorSession;
 import io.trino.spi.connector.ConnectorTableMetadata;
@@ -22,6 +23,7 @@ import io.trino.spi.connector.InMemoryRecordSet;
 import io.trino.spi.connector.SchemaTableName;
 import io.trino.spi.connector.SystemTable;
 import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeManager;
 
 import javax.inject.Inject;
 
@@ -33,7 +35,14 @@ import java.util.stream.IntStream;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
 import static com.google.common.collect.Streams.stream;
+import static io.trino.plugin.hive.HiveSessionProperties.getTimestampPrecision;
 import static io.trino.plugin.hive.SystemTableHandler.PARTITIONS;
+import static io.trino.plugin.hive.metastore.MetastoreUtil.getProtectMode;
+import static io.trino.plugin.hive.metastore.MetastoreUtil.verifyOnline;
+import static io.trino.plugin.hive.util.HiveBucketing.getHiveBucketHandle;
+import static io.trino.plugin.hive.util.HiveUtil.getPartitionKeyColumnHandles;
+import static io.trino.plugin.hive.util.HiveUtil.getRegularColumnHandles;
+import static io.trino.plugin.hive.util.HiveUtil.isDeltaLakeTable;
 import static io.trino.plugin.hive.util.SystemTables.createSystemTable;
 import static java.util.Objects.requireNonNull;
 import static java.util.function.Function.identity;
@@ -43,11 +52,13 @@ public class PartitionsSystemTableProvider
         implements SystemTableProvider
 {
     private final HivePartitionManager partitionManager;
+    private final TypeManager typeManager;
 
     @Inject
-    public PartitionsSystemTableProvider(HivePartitionManager partitionManager)
+    public PartitionsSystemTableProvider(HivePartitionManager partitionManager, TypeManager typeManager)
     {
         this.partitionManager = requireNonNull(partitionManager, "partitionManager is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
     }
 
     @Override
@@ -68,11 +79,20 @@ public class PartitionsSystemTableProvider
         }
 
         SchemaTableName sourceTableName = PARTITIONS.getSourceTableName(tableName);
-        HiveTableHandle sourceTableHandle = metadata.getTableHandle(session, sourceTableName);
-
-        if (sourceTableHandle == null) {
+        Table sourceTable = metadata.getMetastore()
+                .getTable(new HiveIdentity(session), sourceTableName.getSchemaName(), sourceTableName.getTableName())
+                .orElse(null);
+        if (sourceTable == null || isDeltaLakeTable(sourceTable)) {
             return Optional.empty();
         }
+        verifyOnline(sourceTableName, Optional.empty(), getProtectMode(sourceTable), sourceTable.getParameters());
+        HiveTableHandle sourceTableHandle = new HiveTableHandle(
+                sourceTableName.getSchemaName(),
+                sourceTableName.getTableName(),
+                sourceTable.getParameters(),
+                getPartitionKeyColumnHandles(sourceTable, typeManager),
+                getRegularColumnHandles(sourceTable, typeManager, getTimestampPrecision(session)),
+                getHiveBucketHandle(session, sourceTable, typeManager));
 
         List<HiveColumnHandle> partitionColumns = sourceTableHandle.getPartitionColumns();
         if (partitionColumns.isEmpty()) {

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PropertiesSystemTableProvider.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PropertiesSystemTableProvider.java
@@ -34,6 +34,7 @@ import java.util.Optional;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static io.trino.plugin.hive.SystemTableHandler.PROPERTIES;
 import static io.trino.plugin.hive.util.HiveUtil.isDeltaLakeTable;
+import static io.trino.plugin.hive.util.HiveUtil.isIcebergTable;
 import static io.trino.plugin.hive.util.SystemTables.createSystemTable;
 
 public class PropertiesSystemTableProvider
@@ -61,7 +62,7 @@ public class PropertiesSystemTableProvider
                 .getTable(new HiveIdentity(session), sourceTableName.getSchemaName(), sourceTableName.getTableName())
                 .orElseThrow(() -> new TableNotFoundException(tableName));
 
-        if (isDeltaLakeTable(table)) {
+        if (isDeltaLakeTable(table) || isIcebergTable(table)) {
             return Optional.empty();
         }
         Map<String, String> sortedTableParameters = ImmutableSortedMap.copyOf(table.getParameters());

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PropertiesSystemTableProvider.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/PropertiesSystemTableProvider.java
@@ -17,7 +17,6 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSortedMap;
 import io.trino.plugin.hive.authentication.HiveIdentity;
 import io.trino.plugin.hive.metastore.Table;
-import io.trino.spi.TrinoException;
 import io.trino.spi.connector.ColumnMetadata;
 import io.trino.spi.connector.ConnectorSession;
 import io.trino.spi.connector.ConnectorTableMetadata;
@@ -33,11 +32,9 @@ import java.util.Map;
 import java.util.Optional;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
-import static io.trino.plugin.hive.HiveErrorCode.HIVE_UNSUPPORTED_FORMAT;
 import static io.trino.plugin.hive.SystemTableHandler.PROPERTIES;
 import static io.trino.plugin.hive.util.HiveUtil.isDeltaLakeTable;
 import static io.trino.plugin.hive.util.SystemTables.createSystemTable;
-import static java.lang.String.format;
 
 public class PropertiesSystemTableProvider
         implements SystemTableProvider
@@ -65,7 +62,7 @@ public class PropertiesSystemTableProvider
                 .orElseThrow(() -> new TableNotFoundException(tableName));
 
         if (isDeltaLakeTable(table)) {
-            throw new TrinoException(HIVE_UNSUPPORTED_FORMAT, format("Cannot query Delta Lake table '%s'", sourceTableName));
+            return Optional.empty();
         }
         Map<String, String> sortedTableParameters = ImmutableSortedMap.copyOf(table.getParameters());
         List<ColumnMetadata> columns = sortedTableParameters.keySet().stream()

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
@@ -180,6 +180,9 @@ public final class HiveUtil
     public static final String SPARK_TABLE_PROVIDER_KEY = "spark.sql.sources.provider";
     public static final String DELTA_LAKE_PROVIDER = "delta";
 
+    public static final String ICEBERG_TABLE_TYPE_NAME = "table_type";
+    public static final String ICEBERG_TABLE_TYPE_VALUE = "iceberg";
+
     private static final DateTimeFormatter HIVE_DATE_PARSER = ISODateTimeFormat.date().withZoneUTC();
     private static final DateTimeFormatter HIVE_TIMESTAMP_PARSER;
     private static final Field COMPRESSION_CODECS_FIELD;
@@ -191,8 +194,6 @@ public final class HiveUtil
     private static final String BIG_DECIMAL_POSTFIX = "BD";
 
     private static final Splitter COLUMN_NAMES_SPLITTER = Splitter.on(',').trimResults().omitEmptyStrings();
-    private static final String ICEBERG_TABLE_TYPE_NAME = "table_type";
-    private static final String ICEBERG_TABLE_TYPE_VALUE = "iceberg";
 
     static {
         DateTimeParser[] timestampWithoutTimeZoneParser = {

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
@@ -839,7 +839,7 @@ public abstract class AbstractTestHive
                             TupleDomain.all()));
                 },
                 ImmutableSet.of(
-                        new PartitionsSystemTableProvider(partitionManager),
+                        new PartitionsSystemTableProvider(partitionManager, TESTING_TYPE_MANAGER),
                         new PropertiesSystemTableProvider()),
                 (metastore) -> new NoneHiveMaterializedViewMetadata()
                 {
@@ -2899,22 +2899,14 @@ public abstract class AbstractTestHive
                         .hasMessage(format("Cannot query Delta Lake table '%s'", tableName));
             }
 
-            // Verify the hidden `$properties` Delta Lake table handle can't be obtained within the hive connector
-            SchemaTableName propertiesTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$properties", tableName.getTableName()));
+            // Verify the hidden `$properties` and `$partitions` Delta Lake table handle can't be obtained within the hive connector
             try (Transaction transaction = newTransaction()) {
                 ConnectorMetadata metadata = transaction.getMetadata();
                 metadata.beginQuery(session);
-                assertThatThrownBy(() -> metadata.getSystemTable(newSession(), propertiesTableName))
-                        .hasMessage(format("Cannot query Delta Lake table '%s'", tableName));
-            }
-
-            // Verify the hidden `$partitions` Delta Lake table handle can't be obtained within the hive connector
-            SchemaTableName partitionsTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$partitions", tableName.getTableName()));
-            try (Transaction transaction = newTransaction()) {
-                ConnectorMetadata metadata = transaction.getMetadata();
-                metadata.beginQuery(session);
-                assertThatThrownBy(() -> metadata.getSystemTable(newSession(), partitionsTableName))
-                        .hasMessage(format("Cannot query Delta Lake table '%s'", tableName));
+                SchemaTableName propertiesTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$properties", tableName.getTableName()));
+                assertThat(metadata.getSystemTable(newSession(), propertiesTableName)).isEmpty();
+                SchemaTableName partitionsTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$partitions", tableName.getTableName()));
+                assertThat(metadata.getSystemTable(newSession(), partitionsTableName)).isEmpty();
             }
 
             // Assert that table is hidden

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
@@ -260,6 +260,8 @@ import static io.trino.plugin.hive.metastore.StorageFormat.fromHiveStorageFormat
 import static io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.cachingHiveMetastore;
 import static io.trino.plugin.hive.util.HiveBucketing.BucketingVersion.BUCKETING_V1;
 import static io.trino.plugin.hive.util.HiveUtil.DELTA_LAKE_PROVIDER;
+import static io.trino.plugin.hive.util.HiveUtil.ICEBERG_TABLE_TYPE_NAME;
+import static io.trino.plugin.hive.util.HiveUtil.ICEBERG_TABLE_TYPE_VALUE;
 import static io.trino.plugin.hive.util.HiveUtil.SPARK_TABLE_PROVIDER_KEY;
 import static io.trino.plugin.hive.util.HiveUtil.columnExtraInfo;
 import static io.trino.plugin.hive.util.HiveUtil.toPartitionValues;
@@ -2935,6 +2937,56 @@ public abstract class AbstractTestHive
                 // list all columns in a table
                 assertThat(listTableColumns(metadata, session, new SchemaTablePrefix(tableName.getSchemaName(), tableName.getTableName())).keySet())
                         .doesNotContain(tableName);
+            }
+        }
+        finally {
+            // Clean up
+            metastoreClient.dropTable(identity, tableName.getSchemaName(), tableName.getTableName(), true);
+        }
+    }
+
+    @Test
+    public void testDisallowQueryingOfIcebergTables()
+    {
+        ConnectorSession session = newSession();
+        HiveIdentity identity = new HiveIdentity(session);
+        SchemaTableName tableName = temporaryTable("trino_iceberg_table");
+
+        Table.Builder table = Table.builder()
+                .setDatabaseName(tableName.getSchemaName())
+                .setTableName(tableName.getTableName())
+                .setOwner(Optional.of(session.getUser()))
+                .setTableType(MANAGED_TABLE.name())
+                .setPartitionColumns(List.of(new Column("a_partition_column", HIVE_INT, Optional.empty())))
+                .setDataColumns(List.of(new Column("a_column", HIVE_STRING, Optional.empty())))
+                .setParameter(ICEBERG_TABLE_TYPE_NAME, ICEBERG_TABLE_TYPE_VALUE);
+        table.getStorageBuilder()
+                .setStorageFormat(fromHiveStorageFormat(PARQUET))
+                .setLocation(getTableDefaultLocation(
+                        metastoreClient.getDatabase(tableName.getSchemaName()).orElseThrow(),
+                        new HdfsContext(session.getIdentity()),
+                        hdfsEnvironment,
+                        tableName.getSchemaName(),
+                        tableName.getTableName()).toString());
+        metastoreClient.createTable(identity, table.build(), NO_PRIVILEGES);
+
+        try {
+            // Verify that the table was created as a Iceberg table can't be queried in hive
+            try (Transaction transaction = newTransaction()) {
+                ConnectorMetadata metadata = transaction.getMetadata();
+                metadata.beginQuery(session);
+                assertThatThrownBy(() -> getTableHandle(metadata, tableName))
+                        .hasMessage(format("Cannot query Iceberg table '%s'", tableName));
+            }
+
+            // Verify the hidden `$properties` and `$partitions` hive system tables table handle can't be obtained for the Iceberg tables
+            try (Transaction transaction = newTransaction()) {
+                ConnectorMetadata metadata = transaction.getMetadata();
+                metadata.beginQuery(session);
+                SchemaTableName propertiesTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$properties", tableName.getTableName()));
+                assertThat(metadata.getSystemTable(newSession(), propertiesTableName)).isEmpty();
+                SchemaTableName partitionsTableName = new SchemaTableName(tableName.getSchemaName(), format("%s$partitions", tableName.getTableName()));
+                assertThat(metadata.getSystemTable(newSession(), partitionsTableName)).isEmpty();
             }
         }
         finally {

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveFileSystem.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveFileSystem.java
@@ -217,7 +217,7 @@ public abstract class AbstractTestHiveFileSystem
                 new NodeVersion("test_version"),
                 new NoneHiveRedirectionsProvider(),
                 ImmutableSet.of(
-                        new PartitionsSystemTableProvider(hivePartitionManager),
+                        new PartitionsSystemTableProvider(hivePartitionManager, TESTING_TYPE_MANAGER),
                         new PropertiesSystemTableProvider()),
                 new DefaultHiveMaterializedViewMetadataFactory(),
                 SqlStandardAccessControlMetadata::new,

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveInMemoryMetastore.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveInMemoryMetastore.java
@@ -61,4 +61,10 @@ public class TestHiveInMemoryMetastore
     {
         throw new SkipException("not supported");
     }
+
+    @Override
+    public void testDisallowQueryingOfIcebergTables()
+    {
+        throw new SkipException("not supported");
+    }
 }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java
@@ -230,9 +230,9 @@ public class TestHiveRedirectionToIceberg
         String icebergTableName = "iceberg.default." + tableName;
 
         createIcebergTable(icebergTableName, false);
-        onTrino().executeQuery("DROP TABLE " + hiveTableName);
-        assertQueryFailure(() -> onTrino().executeQuery("TABLE " + icebergTableName))
-                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): line 1:1: Table '" + icebergTableName + "' does not exist");
+        //TODO restore test assertions after adding redirection awareness to the DropTableTask
+        assertQueryFailure(() -> onTrino().executeQuery("DROP TABLE " + hiveTableName))
+                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): Cannot query Iceberg table 'default." + tableName + "'");
     }
 
     @Test(groups = {HIVE_ICEBERG_REDIRECTIONS, PROFILE_SPECIFIC_TESTS})
@@ -305,18 +305,11 @@ public class TestHiveRedirectionToIceberg
 
         createIcebergTable(icebergTableName, false);
 
-        onTrino().executeQuery("ALTER TABLE " + hiveTableName + " RENAME TO " + tableName + "_new");
+        //TODO restore test assertions after adding redirection awareness to the RenameTableTask
+        assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE " + hiveTableName + " RENAME TO " + tableName + "_new"))
+                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): Cannot query Iceberg table 'default." + tableName + "'");
 
-        assertQueryFailure(() -> onTrino().executeQuery("TABLE " + hiveTableName))
-                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): line 1:1: Table '" + hiveTableName + "' does not exist");
-        assertQueryFailure(() -> onTrino().executeQuery("TABLE " + icebergTableName))
-                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): line 1:1: Table '" + icebergTableName + "' does not exist");
-
-        assertResultsEqual(
-                onTrino().executeQuery("TABLE " + icebergTableName + "_new"),
-                onTrino().executeQuery("TABLE " + hiveTableName + "_new"));
-
-        onTrino().executeQuery("DROP TABLE " + icebergTableName + "_new");
+        onTrino().executeQuery("DROP TABLE " + icebergTableName);
     }
 
     @Test(groups = {HIVE_ICEBERG_REDIRECTIONS, PROFILE_SPECIFIC_TESTS})
@@ -328,15 +321,9 @@ public class TestHiveRedirectionToIceberg
 
         createIcebergTable(icebergTableName, false);
 
-        onTrino().executeQuery("ALTER TABLE " + hiveTableName + " ADD COLUMN some_new_column double");
-
-        // TODO: ALTER TABLE succeeded, but new column was not added
-        Assertions.assertThat(onTrino().executeQuery("DESCRIBE " + icebergTableName).column(1))
-                .containsOnly("nationkey", "name", "regionkey", "comment");
-
-        assertResultsEqual(
-                onTrino().executeQuery("TABLE " + icebergTableName),
-                onTrino().executeQuery("SELECT * /*, NULL*/ FROM tpch.tiny.nation"));
+        //TODO restore test assertions after adding redirection awareness to the AddColumnTask
+        assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE " + hiveTableName + " ADD COLUMN some_new_column double"))
+                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): Cannot query Iceberg table 'default." + tableName + "'");
 
         onTrino().executeQuery("DROP TABLE " + icebergTableName);
     }
@@ -353,11 +340,9 @@ public class TestHiveRedirectionToIceberg
         assertTableComment("hive", "default", tableName).isNull();
         assertTableComment("iceberg", "default", tableName).isNull();
 
-        onTrino().executeQuery("COMMENT ON TABLE " + hiveTableName + " IS 'This is my table, there are many like it but this one is mine'");
-
-        // TODO: COMMENT ON TABLE succeeded, but comment was not preserved
-        assertTableComment("hive", "default", tableName).isNull();
-        assertTableComment("iceberg", "default", tableName).isNull();
+        //TODO restore test assertions after adding redirection awareness to the CommentTask
+        assertQueryFailure(() -> onTrino().executeQuery("COMMENT ON TABLE " + hiveTableName + " IS 'This is my table, there are many like it but this one is mine'"))
+                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): Cannot query Iceberg table 'default." + tableName + "'");
 
         onTrino().executeQuery("DROP TABLE " + icebergTableName);
     }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergHiveTablesCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergHiveTablesCompatibility.java
@@ -17,12 +17,12 @@ import io.trino.tempto.ProductTest;
 import org.testng.annotations.Test;
 
 import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure;
-import static io.trino.tempto.assertions.QueryAssert.assertThat;
 import static io.trino.tests.product.TestGroups.HMS_ONLY;
 import static io.trino.tests.product.TestGroups.ICEBERG;
 import static io.trino.tests.product.TestGroups.STORAGE_FORMATS;
 import static io.trino.tests.product.hive.util.TemporaryHiveTable.randomTableSuffix;
 import static io.trino.tests.product.utils.QueryExecutors.onTrino;
+import static java.lang.String.format;
 
 public class TestIcebergHiveTablesCompatibility
         extends ProductTest
@@ -49,15 +49,13 @@ public class TestIcebergHiveTablesCompatibility
         onTrino().executeQuery("CREATE TABLE iceberg.default." + tableName + "(a bigint)");
 
         assertQueryFailure(() -> onTrino().executeQuery("SELECT * FROM hive.default." + tableName))
-                // TODO (https://github.com/trinodb/trino/issues/8693) throw specific exception message
-                .hasMessageMatching("Query failed \\(#\\w+\\):\\Q Unable to create input format org.apache.hadoop.mapred.FileInputFormat");
+                .hasMessageMatching(format("Query failed \\(#\\w+\\):\\Q Cannot query Iceberg table 'default.%s'", tableName));
 
         assertQueryFailure(() -> onTrino().executeQuery("SELECT * FROM hive.default.\"" + tableName + "$partitions\""))
                 .hasMessageMatching("Query failed \\(#\\w+\\):\\Q line 1:15: Table 'hive.default." + tableName + "$partitions' does not exist");
 
-        // TODO (https://github.com/trinodb/trino/issues/8693) should fail
-        assertThat(onTrino().executeQuery("SELECT * FROM hive.default.\"" + tableName + "$properties\""))
-                .hasRowsCount(1);
+        assertQueryFailure(() -> onTrino().executeQuery("SELECT * FROM hive.default.\"" + tableName + "$properties\""))
+                .hasMessageMatching("Query failed \\(#\\w+\\):\\Q line 1:15: Table 'hive.default." + tableName + "$properties' does not exist");
 
         onTrino().executeQuery("DROP TABLE iceberg.default." + tableName);
     }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergHiveViewsCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergHiveViewsCompatibility.java
@@ -21,6 +21,7 @@ import org.testng.annotations.Test;
 import java.util.List;
 
 import static io.trino.tempto.assertions.QueryAssert.Row.row;
+import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure;
 import static io.trino.tempto.assertions.QueryAssert.assertThat;
 import static io.trino.tests.product.TestGroups.HMS_ONLY;
 import static io.trino.tests.product.TestGroups.ICEBERG;
@@ -53,8 +54,8 @@ public class TestIcebergHiveViewsCompatibility
             onTrino().executeQuery("CREATE VIEW hive.default.hive_view_qualified_hive AS SELECT * FROM hive.default.hive_table");
             onTrino().executeQuery("CREATE VIEW hive.default.hive_view_unqualified_hive AS SELECT * FROM hive_table");
             onTrino().executeQuery("CREATE VIEW hive.default.hive_view_qualified_iceberg AS SELECT * FROM iceberg.default.iceberg_table");
-            // this should probably fail but it does not now; testing current behavior as a documentation
-            onTrino().executeQuery("CREATE VIEW hive.default.hive_view_unqualified_iceberg AS SELECT * FROM iceberg_table");
+            assertQueryFailure(() -> onTrino().executeQuery("CREATE VIEW hive.default.hive_view_unqualified_iceberg AS SELECT * FROM iceberg_table"))
+                    .hasMessageMatching("Query failed \\(#\\w+\\):\\Q Cannot query Iceberg table 'default.iceberg_table'");
 
             onTrino().executeQuery("USE iceberg.default"); // for sake of unqualified table references
             onTrino().executeQuery("CREATE VIEW iceberg.default.iceberg_view_qualified_hive AS SELECT * FROM hive.default.hive_table");
@@ -75,7 +76,6 @@ public class TestIcebergHiveViewsCompatibility
                             .add(row("hive_view_qualified_hive"))
                             .add(row("hive_view_unqualified_hive"))
                             .add(row("hive_view_qualified_iceberg"))
-                            .add(row("hive_view_unqualified_iceberg"))
                             .add(row("iceberg_view_qualified_hive"))
                             .add(row("iceberg_view_qualified_iceberg"))
                             .add(row("iceberg_view_unqualified_iceberg"))
@@ -88,7 +88,6 @@ public class TestIcebergHiveViewsCompatibility
                             .add(row("hive_view_qualified_hive"))
                             .add(row("hive_view_unqualified_hive"))
                             .add(row("hive_view_qualified_iceberg"))
-                            .add(row("hive_view_unqualified_iceberg"))
                             .add(row("iceberg_view_qualified_hive"))
                             .add(row("iceberg_view_qualified_iceberg"))
                             .add(row("iceberg_view_unqualified_iceberg"))
@@ -98,10 +97,6 @@ public class TestIcebergHiveViewsCompatibility
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.hive_view_qualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.hive_view_unqualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.hive_view_qualified_iceberg")).containsOnly(row(2));
-            assertThatThrownBy(() -> onTrino().executeQuery("SELECT * FROM hive.default.hive_view_unqualified_iceberg"))
-                    // hive connector tries to read from iceberg table
-                    // TODO: make query fail with nicer message
-                    .hasMessageContaining("Unable to create input format org.apache.hadoop.mapred.FileInputFormat");
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.iceberg_view_qualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.iceberg_view_qualified_iceberg")).containsOnly(row(2));
             assertThat(onTrino().executeQuery("SELECT * FROM hive.default.iceberg_view_unqualified_iceberg")).containsOnly(row(2));
@@ -110,10 +105,6 @@ public class TestIcebergHiveViewsCompatibility
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.hive_view_qualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.hive_view_unqualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.hive_view_qualified_iceberg")).containsOnly(row(2));
-            assertThatThrownBy(() -> onTrino().executeQuery("SELECT * FROM iceberg.default.hive_view_unqualified_iceberg"))
-                    // hive connector tries to read from iceberg table
-                    // TODO: make query fail with nicer message
-                    .hasMessageContaining("Unable to create input format org.apache.hadoop.mapred.FileInputFormat");
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.iceberg_view_qualified_hive")).containsOnly(row(1));
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.iceberg_view_qualified_iceberg")).containsOnly(row(2));
             assertThat(onTrino().executeQuery("SELECT * FROM iceberg.default.iceberg_view_unqualified_iceberg")).containsOnly(row(2));

